torch/distributions/mixture_same_family.py

```diff
@@ -0,0 +1,181 @@
+
+    Examples::
+
+        # Construct Gaussian Mixture Model in 1D consisting of 5 equally
+        # weighted normal distributions
+        >>> mix = D.Categorical(torch.ones(5,))
+        >>> comp = D.Normal(torch.randn(5,), torch.rand(5,))
+        >>> gmm = MixtureSameFamily(mix, comp)
+
+        # Construct Gaussian Mixture Modle in 2D consisting of 5 equally
+        # weighted bivariate normal distributions
+        >>> mix = D.Categorical(torch.ones(5,))
+        >>> comp = D.Independent(D.Normal(
+                     torch.randn(5,2), torch.rand(5,2)), 1)
+        >>> gmm = MixtureSameFamily(mix, comp)
+
+        # Construct a batch of 3 Gaussian Mixture Models in 2D each
+        # consisting of 5 random weighted bivariate normal distributions
+        >>> mix = D.Categorical(torch.rand(3,5))
+        >>> comp = D.Independent(D.Normal(
+                    torch.randn(3,5,2), torch.rand(3,5,2)), 1)
+        >>> gmm = MixtureSameFamily(mix, comp)
+
+    Args:
+        mixture_distribution: `torch.distributions.Categorical`-like
+            instance. Manages the probability of selecting component.
+            The number of categories must match the rightmost batch
+            dimension of the `component_distribution`. Must have either
+            scalar `batch_shape` or `batch_shape` matching
+            `component_distribution.batch_shape[:-1]`
+        component_distribution: `torch.distributions.Distribution`-like
+            instance. Right-most batch dimension indexes component.
+    """
+    arg_constraints = {}
+    has_rsample = False
+
+    def __init__(self,
+                 mixture_distribution,
+                 component_distribution,
+                 validate_args=None):
+        self._mixture_distribution = mixture_distribution
+        self._component_distribution = component_distribution
+
+        if not isinstance(self._mixture_distribution, Categorical):
+            raise ValueError(" The Mixture distribution needs to be an "
+                             " instance of torch.distribtutions.Categorical")
+
+        if not isinstance(self._component_distribution, Distribution):
+            raise ValueError("The Component distribution need to be an "
+                             "instance of torch.distributions.Distribution")
+
+        # Check that batch size matches
+        mdbs = self._mixture_distribution.batch_shape
+        cdbs = self._component_distribution.batch_shape[:-1]
+        if len(mdbs) != 0 and mdbs != cdbs:
+            raise ValueError("`mixture_distribution.batch_shape` ({0}) is not "
+                             "compatible with `component_distribution."
+                             "batch_shape`({1})".format(mdbs, cdbs))
+
+        # Check that the number of mixture component matches
+        km = self._mixture_distribution.logits.shape[-1]
+        kc = self._component_distribution.batch_shape[-1]
+        if km is not None and kc is not None and km != kc:
```

Comment:
I think this can be stricter:
```py
if mixture_distribution.event_shape != component_distribution.batch_shape[-1:]:
    raise ...
```

Reply:
I am not sure about this. It implies that the number of components can span multiple dimensions, instead of a single one. Not sure that the rest of the code support this.

Reply:
Hmm, how about additionally checking that `len(mixture_distribution.event_shape) == 1`? I just don't see any checks that the `mixture_distribution` is a `Categorical`, and I don't see how an error would be raised if a user tried to pass e.g. a `Bernoulli` distribution.

Reply:
Checking that the `mixture_distribution` is `Categorical` is done as the very first thing:  
https://github.com/SkafteNicki/pytorch/blob/909c4d3849dfcc6bfbbc577843917d8a6768ac08/torch/distributions/mixture_same_family.py#L58-L60